independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/LocalProject.java

```diff
@@ -228,20 +230,28 @@ public Path getSourcesDir() {
         return getSourcesSourcesDir().getParent();
     }
 
-    public Path getResourcesSourcesDir() {
+    public Set<Path> getResourcesSourcesDirs() {
```

Comment:
About using `Set`s here and a few other places that represent resources of a single module, these paths will be used to compose a classpath, so it make it consistent, shouldn't we always use `List`s instead? Is there really a high chance of finding duplicates in it?

Reply:
I can change it for `List` everywhere if needed but I decided to use `Set` instead because:

1. I realized that in many places in this part of the code, `Set`s are used when a collection is needed like for example in `SourceSet`/`SourceSetImpl` or `sourcePaths` in `DevMode`
2. We could have duplicates for source directories for example when we use filtering and we would like to enable filtering for specific files only like [this example](https://stackoverflow.com/a/46233958/1997376)

Reply:
Now I'm wondering why we are using `Set<File>` in `SourceSet` when we already have https://github.com/quarkusio/quarkus/blob/main/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/bootstrap/model/PathsCollection.java
Which could be enhanced to not add duplicates. While duplicate paths are possible I guess that'll still be rear. Another guess is that it's also unlikely we'll have to deal with more than a few resources dirs, in which case lists are more efficient than sets even for `contains(item)`.
@stuartwdouglas do you see a problem switching to a list-based collection?

Reply:
I think we should just define `PathsCollection` as an ordered collection of paths that does not contain duplicates and use it everywhere.

Reply:
It should probably also be renamed to `PathCollection`.

Reply:
+1 to making this a list, we really don't want non-deterministic behavior here.

Reply:
OK I change it for PathCollection then

Reply:
@aloubyansky do I include the renaming of `PathsCollection` in this PR?

Reply:
I think let's do that later. Thanks!

Reply:
@aloubyansky Just to be sure, you want `PathsCollection` to be used everywhere we have a Set of paths? Like in CompilationUnit, in DevModeContext and so on, or only in `SourceSet`?

Reply:
That'd be a massive change probably but I guess it makes sense. Unless @stuartwdouglas thinks it's too far, I'd say let's do that.

Reply:
@aloubyansky Unfortunately there are several places in the code that expect a mutable collection (like for example [this line](https://github.com/quarkusio/quarkus/blob/main/devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java#L226) and [this method](https://github.com/quarkusio/quarkus/blob/main/core/deployment/src/main/java/io/quarkus/deployment/dev/DevModeContext.java#L271-L275)). So it sounds like more complex than we thought. What do you want me to do? Add a mutable counterpart of `PathsCollection`  and use it only where it is needed, something else?

Reply:
> @aloubyansky Unfortunately there are several places in the code that expect a mutable collection (like for example [this line](https://github.com/quarkusio/quarkus/blob/main/devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java#L226) and [this method]

This one looks easy to workaround. Basically, it a WorkspaceModuleImpl there. So you could use PathsCollection.builder() adding both source main and test source sets to it. The build it and create the WorkspaceModule. Right?

> (https://github.com/quarkusio/quarkus/blob/main/core/deployment/src/main/java/io/quarkus/deployment/dev/DevModeContext.java#L271-L275)).

This one, purely from the API design perspective, doesn't look right. But it's better not to touch it as part of this PR. What you could do is `PathsCollection.from(sourcePaths)` then add the other paths and `build()`. Or add `add(Iterable<Path> paths)` to `PathsCollection` in a similar way to `add(Path... paths)`.

Reply:
I am happy with moving everything, although TBH personally I would keep this PR narrow in scope and then we can do more changes after it is merged.

Reply:
> I am happy with moving everything, although TBH personally I would keep this PR narrow in scope and then we can do more changes after it is merged.

@aloubyansky Considering the remark of @stuartwdouglas, do I revert all the changes related to `PathsCollection` ?

Reply:
It looks like they are in a separate commit, right? If so then, IMO, it should be ok in case we want to revert it for some reason after the merge.

Reply:
Yes I separated the commits intentionally as I anticipated a little bit this possibility

Reply:
If you have it all working then I am fine with it, I didn't look at the current state before I commented.

Reply:
Because of the run task of Gradle, the methods of `SourceSet` need to return interface types to be able to create proxies on top of it otherwise you get error of the next type. I think that I have no other choices but to convert `PathsCollection` into an interface, Is that ok with you?
```
2021-05-18T11:24:09.3755726Z     Exception in thread "main" java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
2021-05-18T11:24:09.3757359Z     	at io.quarkus.launcher.QuarkusLauncher.launch(QuarkusLauncher.java:58)
2021-05-18T11:24:09.3758732Z     	at io.quarkus.runtime.Quarkus.launchFromIDE(Quarkus.java:96)
2021-05-18T11:24:09.3759816Z     	at io.quarkus.runtime.Quarkus.run(Quarkus.java:83)
2021-05-18T11:24:09.3761347Z     	at io.quarkus.runtime.Quarkus.run(Quarkus.java:42)
2021-05-18T11:24:09.3762280Z     	at io.quarkus.runtime.Quarkus.run(Quarkus.java:119)
2021-05-18T11:24:09.3763157Z     	at org.acme.EntryPoint.main(EntryPoint.java:9)
2021-05-18T11:24:09.3764548Z     Caused by: java.lang.reflect.InvocationTargetException
2021-05-18T11:24:09.3766148Z     	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2021-05-18T11:24:09.3768290Z     	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2021-05-18T11:24:09.3770650Z     	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2021-05-18T11:24:09.3772555Z     	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2021-05-18T11:24:09.3773843Z     	at io.quarkus.launcher.QuarkusLauncher.launch(QuarkusLauncher.java:56)
2021-05-18T11:24:09.3775191Z     	... 5 more
2021-05-18T11:24:10.1218864Z     Caused by: java.lang.RuntimeException: java.lang.IllegalArgumentException: io.quarkus.bootstrap.model.PathsCollection is not an interface
2021-05-18T11:24:10.1234536Z     	at io.quarkus.bootstrap.IDELauncherImpl.launch(IDELauncherImpl.java:85)
2021-05-18T11:24:10.1235855Z     	... 10 more
2021-05-18T11:24:10.1237283Z     Caused by: java.lang.IllegalArgumentException: io.quarkus.bootstrap.model.PathsCollection is not an interface
2021-05-18T11:24:10.1239583Z     	at java.base/java.lang.reflect.Proxy$ProxyBuilder.validateProxyInterfaces(Proxy.java:688)
2021-05-18T11:24:10.1241123Z     	at java.base/java.lang.reflect.Proxy$ProxyBuilder.<init>(Proxy.java:627)
2021-05-18T11:24:10.1242321Z     	at java.base/java.lang.reflect.Proxy$ProxyBuilder.<init>(Proxy.java:635)
2021-05-18T11:24:10.1243802Z     	at java.base/java.lang.reflect.Proxy.lambda$getProxyConstructor$0(Proxy.java:415)
2021-05-18T11:24:10.1249164Z     	at java.base/jdk.internal.loader.AbstractClassLoaderValue$Memoizer.get(AbstractClassLoaderValue.java:329)
2021-05-18T11:24:10.1252515Z     	at java.base/jdk.internal.loader.AbstractClassLoaderValue.computeIfAbsent(AbstractClassLoaderValue.java:205)
2021-05-18T11:24:10.1258098Z     	at java.base/java.lang.reflect.Proxy.getProxyConstructor(Proxy.java:413)
2021-05-18T11:24:10.1260384Z     	at java.base/java.lang.reflect.Proxy.newProxyInstance(Proxy.java:1006)
2021-05-18T11:24:10.1264309Z     	at org.gradle.tooling.internal.adapter.ProtocolToModelAdapter.createView(ProtocolToModelAdapter.java:160)
2021-05-18T11:24:10.1267630Z     	at org.gradle.tooling.internal.adapter.ProtocolToModelAdapter.convert(ProtocolToModelAdapter.java:276)
2021-05-18T11:24:10.1280517Z     	at org.gradle.tooling.internal.adapter.ProtocolToModelAdapter.access$1500(ProtocolToModelAdapter.java:56)
2021-05-18T11:24:10.1283751Z     	at org.gradle.tooling.internal.adapter.ProtocolToModelAdapter$AdaptingMethodInvoker.invoke(ProtocolToModelAdapter.java:477)
2021-05-18T11:24:10.1287361Z     	at org.gradle.tooling.internal.adapter.ProtocolToModelAdapter$PropertyCachingMethodInvoker.invoke(ProtocolToModelAdapter.java:705)
2021-05-18T11:24:10.1290567Z     	at org.gradle.tooling.internal.adapter.ProtocolToModelAdapter$SafeMethodInvoker.invoke(ProtocolToModelAdapter.java:742)
2021-05-18T11:24:10.1305810Z     	at org.gradle.tooling.internal.adapter.ProtocolToModelAdapter$SupportedPropertyInvoker.invoke(ProtocolToModelAdapter.java:766)
2021-05-18T11:24:10.1309379Z     	at org.gradle.tooling.internal.adapter.ProtocolToModelAdapter$InvocationHandlerImpl.invoke(ProtocolToModelAdapter.java:432)
2021-05-18T11:24:10.1311418Z     	at com.sun.proxy.$Proxy17.getSourceDirectories(Unknown Source)
2021-05-18T11:24:10.1312963Z     	at io.quarkus.bootstrap.util.QuarkusModelHelper.convert(QuarkusModelHelper.java:147)
2021-05-18T11:24:10.1315315Z     	at io.quarkus.bootstrap.util.QuarkusModelHelper.serializeAppModel(QuarkusModelHelper.java:62)
2021-05-18T11:24:10.1317757Z     	at io.quarkus.bootstrap.util.QuarkusModelHelper.exportModel(QuarkusModelHelper.java:50)
2021-05-18T11:24:10.1353249Z     	at io.quarkus.bootstrap.utils.BuildToolHelper.enableGradleAppModelForDevMode(BuildToolHelper.java:120)
2021-05-18T11:24:10.1355698Z     	at io.quarkus.bootstrap.IDELauncherImpl.launch(IDELauncherImpl.java:41)
```

[The corresponding logs](https://pipelines.actions.githubusercontent.com/5Na9LjmLgXLiLuzJ0Z1qHRbIDVUXUHNrKe6jPcqGdw5t1S8TiW/_apis/pipelines/1/runs/118987/signedlogcontent/48?urlExpires=2021-05-18T14%3A03%3A48.6088018Z&urlSigningMethod=HMACV1&urlSignature=vaLsuyzqJcfrjib5ic8f41RXAAOBL0%2BPbtNVuVjEcAo%3D)

Reply:
Oh, right. Yes.